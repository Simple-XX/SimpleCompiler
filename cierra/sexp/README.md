# Sexp

This is an auxiliary library for generating smt-lib v2 flavored S-expressions.

Most part of its code is adopted from [rust-sise](https://github.com/eduardosm/rust-sise).

The modified version is licensed under either of Apache License, Version 2.0 or MIT license, just like the original one.