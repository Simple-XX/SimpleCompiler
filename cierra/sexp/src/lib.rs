#![allow(clippy::missing_panics_doc, clippy::module_name_repetitions)]

pub use serializer::{Serializer, SerializerStyle};

mod serializer;
mod utils;

pub trait SexpDisplay {
    fn fmt(&self, s: &mut Serializer);
    fn display(&self, style: SerializerStyle) -> String {
        let mut buf = String::new();
        let mut s = Serializer::new(style, &mut buf);
        self.fmt(&mut s);
        s.finish(true);
        buf
    }
    fn pretty_display(&self) -> String {
        self.display(SerializerStyle { line_break: "\n", indentation: "  " })
    }
    fn plain_display(&self) -> String {
        self.display(SerializerStyle { line_break: "", indentation: "" })
    }
}

impl<T> SexpDisplay for &T
where
    T: SexpDisplay,
{
    fn fmt(&self, s: &mut Serializer) {
        SexpDisplay::fmt(&**self, s);
    }
}

impl<T> SexpDisplay for &mut T
where
    T: SexpDisplay,
{
    fn fmt(&self, s: &mut Serializer) {
        SexpDisplay::fmt(&**self, s);
    }
}

impl<T> SexpDisplay for Box<T>
where
    T: SexpDisplay,
{
    fn fmt(&self, s: &mut Serializer) {
        SexpDisplay::fmt(&**self, s);
    }
}

impl SexpDisplay for &dyn SexpDisplay {
    fn fmt(&self, s: &mut Serializer) {
        SexpDisplay::fmt(&**self, s);
    }
}

impl SexpDisplay for Box<dyn SexpDisplay> {
    fn fmt(&self, s: &mut Serializer) {
        SexpDisplay::fmt(&**self, s);
    }
}
