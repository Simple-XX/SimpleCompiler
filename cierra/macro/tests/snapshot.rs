use insta::assert_display_snapshot;
use sexp::{Serializer, SerializerStyle, SexpDisplay};
use sexp_macro::sexp;

struct Plain<F>(F)
where
    F: FnOnce(&mut Serializer);

impl<F> SexpDisplay for Plain<F>
where
    F: FnOnce(&mut Serializer),
{
    fn fmt(&self, s: &mut Serializer) {
        sexp!(s, (assert 1 2));
    }
}

macro_rules! plain {
    ($($tt: tt)*) => {
        Plain(|s| {sexp!(s, $($tt)*);})
    }
}

#[test]
fn snapshot() {
    snapshot_sexp("simple", |s| {
        sexp!(s, (assert 1 2));
    });
    snapshot_sexp("symbol_ex", |s| {
        sexp!(s, (= y (+ x 1)));
    });
    let p = plain!((not (= x 0)));
    let q = plain!((> y 1));
    snapshot_sexp("nested", |s| {
        sexp!(s, (=> %p %q));
    });
    snapshot_sexp("bracket_1", |s| {
        sexp!(s, (forall [x y z] (=> (and (<= x y) (<= y z)) (<= x z))));
    });
    snapshot_sexp("bracket_2", |s| {
        sexp!(s, (forall [x] (= x x)));
    });
    let l: [Box<dyn SexpDisplay>; 2] = [Box::new(plain!((not (= x 0)))), Box::new(plain!((> y 1)))];
    snapshot_sexp("expand", |s| {
        sexp!(s, (test ...(&l)));
    });
    snapshot_sexp("long", |s| {
        sexp!(s, (assert 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20));
    });
    snapshot_sexp("long_nested", |s| {
        sexp!(s, (assert (not (=> (choice (+ (choice (8 9 10)) (assume (- (choice (8 (choice (8 9 10)) (choice (+ (assume (- (choice (8 (choice (+ (assume (- (choice (8 9 10)))))) 10)))))) 10))))))))));
    });
}

fn snapshot_sexp(name: &str, f: impl FnOnce(&mut Serializer)) {
    let mut buf = String::new();
    let mut s = Serializer::new(SerializerStyle { line_break: "\n", indentation: "  " }, &mut buf);
    f(&mut s);
    s.finish(true);
    assert_display_snapshot!(name, buf);
}
