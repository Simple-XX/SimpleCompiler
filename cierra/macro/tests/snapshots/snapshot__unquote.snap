---
source: macro/tests/snapshot.rs
expression: buf
---
(=>
  (not
    (= x 0)
  )
  (> y 1)
)

